### PR TITLE
Added update Meta Data cache after each change to meta_data array

### PR DESCRIPTION
**Pull request by alarocca130**
**tl;dr**
meta_data cache, set in _includes/abstracts/_**abstract-wc-data.php**:441needs to be updated each time a method changes meta array

I added `update_meta_data_cache()` method to `WC_data` class, to perform this operation

# Problem Description

It seems that if we use any of the `WC_Data`'s method to update meta_data array, and then we get a new instance of the same `WC_Data `object, changes to meta data array are lost

Let's make a proper example, considering a Product (`WC_Product` extends `WC_Data`)
If I get the product object, using `wc_get_product()` function, and I add some meta data to it, using `add_meta_data()` method, everything seems to work just fine

Anyway, if then I get the same product, calling `wc_get_product()` again, my changes to meta data are lost

This could be totally ok, by the way, so please ignore this request if this was intended; anyway I think that product instances should match each other, especially when talking about meta data, that could be stored into DB in a second moment

# How to recreate the problem

Using latest WooCommerce commit, simply follow these steps

1. Add the following code to **functions.php** file of your theme or child
```
function test1(){
    $product = wc_get_product(70);
    $product->add_meta_data( 'test_key', 'test_value' );
}
add_action( 'init', 'test1' );

function test2(){
	$product = wc_get_product(70);
	var_dump( $product->get_meta( 'test_key', true ) );die();
}
add_action( 'init', 'test2', 20 );
```
2. Visit any page of the site

My code just performs two different actions on `init` hook:
- first (priority 10) set a meta for the product with id 70 (change that to any product ID actually present in your DB)
- then (priority 20) retrieves back the product with id 70, and print value for the meta added before

This code should just print an empty string, to demonstrate that my meta is lost

# Why this is happening

I think that the problem is related to the fact that cache for `meta_data` is set on the first call of `read_meta_data()` method, using values read from DB, but it never gets updated

Once cache is set, each subsequent call to `read_meta_data()` will just return the meta cache, that could, at this point, contain outdated informations

# Possible solution to the problem

An easy fix would be to update cached values each time something changes into `meta_data` array

I added the `update_meta_data_cache()` method to `WC_Data` class, in order to perform this operation, and I simply call it each time a change is performed over `meta_data` property
This means in the following methods:
- `add_meta_data()`
- `update_meta_data()`
- `delete_meta_data()`

### CODE REVIEW

**Comment by kreamweb** [comment]
I have the same issue with my plugins, the table of postmeta has several rows with the same meta_key! 
And the changes are not recovered!  :(

**Comment by andreagrillo** [comment]
me, too! Same issue!

**Comment by mikejolley** [comment]
You're missing a vital piece; meta changes are not persisted until SAVE.

Add meta data adds to the internal meta array, not the DB: https://github.com/woocommerce/woocommerce/blob/master/includes/abstracts/abstract-wc-data.php#L341-L350

Same for update and delete.

Your workaround doesn't actually save anything to the database, it just updates the cache with meta values which will eventually be lost when the cache is invalidated (unless something saves).

The correct gist would be:

```
function test1(){
    $product = wc_get_product(70);
    $product->add_meta_data( 'test_key', 'test_value' );
    $product->save(); // Persist changes and clears caches.
}
add_action( 'init', 'test1' );

function test2(){
	$product = wc_get_product(70);
	var_dump( $product->get_meta( 'test_key', true ) );die();
}
add_action( 'init', 'test2', 20 );
```

This is all by design and is not meant to mirror WordPress' 'update now right to the database' functions.

**Comment by andreagrillo** [comment]
Hi, 

ok, I can save it but in this way I don't use the cached value but rea the information from DB. I think this Impact on performance. In my personal opinion it makes no sense to use a cache system in this way because I haven't any benefit if I always have to read/write from/to database

**Comment by mikejolley** [comment]
The cache system is used if you read the same object multiple times. Not all interactions with objects are for the purposes of saving.

**Comment by andreagrillo** [comment]
I'm sorry to bother you again but  if possible pleaes pay attention to this scenario: 

If you have two different actions in the same execution and you can't access directly to product object you need to get this using `wc_get_product(...)`  

This functions create a new instance of WC_Product obejct (new wc_product()) that is different then the previous one...

Please, can you tell me what is your opinion about that ?

**Comment by mikejolley** [comment]
Thats how it should be, if you didn't save the first object.
